From 21ee75079ec354e2e5ba3252cdc63be4da059413 Mon Sep 17 00:00:00 2001
From: Andrew Walker <awalker@ixsystems.com>
Date: Tue, 9 Nov 2021 13:46:45 -0500
Subject: [PATCH] s3/winbindd/winbindd_util - fix "allow trusted domains"

At bypass for BUILTIN (S-1-5-32) domain if
"allow trusted domains" is disabled.
---
 source3/winbindd/winbindd_util.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/source3/winbindd/winbindd_util.c b/source3/winbindd/winbindd_util.c
index 1ae4a8d3ca3..20f13fcaa21 100644
--- a/source3/winbindd/winbindd_util.c
+++ b/source3/winbindd/winbindd_util.c
@@ -125,13 +125,19 @@ static NTSTATUS add_trusted_domain(const char *domain_name,
 	struct winbindd_domain *domain = NULL;
 	int role = lp_server_role();
 	struct dom_sid_buf buf;
+	bool is_builtin = false;
 
 	if (is_null_sid(sid)) {
 		DBG_ERR("Got null SID for domain [%s]\n", domain_name);
 		return NT_STATUS_INVALID_PARAMETER;
 	}
 
-	if (!is_allowed_domain(domain_name)) {
+	if (strequal(domain_name, "BUILTIN") &&
+	    sid_check_is_builtin(sid)) {
+		is_builtin = True;
+	}
+
+	if (!is_builtin && !is_allowed_domain(domain_name)) {
 		return NT_STATUS_NO_SUCH_DOMAIN;
 	}
 
-- 
2.26.2

